### DISCUSSION TITLE
Checks the access code used to arm or disarm

### DISCUSSION THREAD

**Discussion by SebastianJoan** [taligentx/dscKeybusInterface#276] [dscKeybusInterface Discussion]
when i try to Checks the access code used to arm or disarm it only show access code used to arm the alarm but it didn´t show the user code used to disarm the alarm 

 if (dsc.accessCodeChanged[partition]) {
        dsc.accessCodeChanged[partition] = false;  // Resets the access code status flag
        Serial.print(F("Partition "));
        Serial.print(partition + 1);
        switch (dsc.accessCode[partition]) {
          case 40: Serial.print(F(": Master")); break;
          default: Serial.print(F(": Access")); break;
        }
        Serial.print(F(" code "));
        Serial.println(dsc.accessCode[partition]);
      }

**Comment by UltimatPronin** [comment]
I confirm the problem

**Comment by taligentx** [comment]
Hi @SebastianJoan @UltimatPronin - the library keeps track of access codes for each partition both for arming and disarming, at any time you can get the access code used on the partition from `dsc.accessCode[partition]`.

However, `dsc.accessCodeChanged[partition]` will only be true if the access code used to arm/disarm the partition is different than the last arm/disarm code.  

If you would like to get the access code used to disarm a partition even if it's the same code used to arm the partition, you can check `dsc.accessCode[partition]` in your sketch, there's no need to check `dsc.accessCodeChanged[partition]`.

Example for the `Status` sketch:
```
Before:
...

        else {
          Serial.print(F("Partition "));
          Serial.print(partition + 1);
          Serial.println(F(": Disarmed"));
        }

After:
...
        else {
          Serial.print(F("Partition "));
          Serial.print(partition + 1);
          Serial.println(F(": Disarmed"));
          Serial.print(F("Partition "));
          Serial.print(partition + 1);
          switch (dsc.accessCode[partition]) {
            case 40: Serial.print(F(": Master")); break;
            default: Serial.print(F(": Access")); break;
          }
          Serial.print(F(" code "));
          Serial.println(dsc.accessCode[partition]);
        }
```
This should be repeated for each section of the sketch that shows when the partition is disarmed.

**Reply by UltimatPronin** [reply]
@taligentx , thank you for reply! 
Works only for 1st time:

For tests I use this:
if (CMD == "ARM"){
          if(!dsc.armed[partition] && !dsc.exitDelay[partition]){ //arm away
            const char* accessCode = "1111"; //user #1 arm  away
            dsc.write(accessCode);    
                       
          }
        }
        else if (CMD == "DISARM"){
          if (dsc.armed[partition] || dsc.exitDelay[partition]) //disarm        
          {
            const char* accessCode = "2222"; //user #2 disarm panel
            dsc.write(accessCode); 
                                    
          }
        }

And then:
      // Checks armed status
      if (dsc.armedChanged[partition]) {
        if (dsc.armed[partition]) {
 
          switch (dsc.accessCode[partition]) {
            case 0:Log_println(F(": User 1")); break;
            case 1:Log_println(F(": User 2")); break;
            //...
            case 40: Log_println(F(": Master")); break;
            default: Log_println(F(": Access")); break;
          } 
         
        }
        else {
            
            switch (dsc.accessCode[partition]) {
              case 0:Log_println(F(": User 1")); break;
              case 1:Log_println(F(": User 2")); break;
              //...
              case 40: Log_println(F(": Master")); break;
              default: Log_println(F(": Access")); break;
            }
             
        }
      }

      if (dsc.armedChanged[partition]) dsc.armedChanged[partition] = false;  // Resets the partition armed status flag

For 1st arm with user 1 access code - I get "armed by user 1"
Then for 1st disarm with user 2 access code - I get "disarmed by user 2"
After this, with many arm/disarm attemps (with CMD == "ARM"/"DISARM") & get only armed/disarmed by user 2, not more by user 1.

Some more investigations:
looks like access code used for disarm stored in dsc.accessCode[partition] until another access code used for disarm.
I tried to use real keypad too - the same problem.

Any ideas? Something more is wrong?

**Comment by taligentx** [comment]
Thanks for the info!  I was premature in converting this issue to a discussion topic, there was a bug where the code that handles the disarm access code was being ignored - fixed in `develop` in #284.  When disarming, the access code in `dsc.accessCode[partition]` should now always be the access code used for disarming.

For arming, there is a small delay (~130ms) from when the panel reports being armed and the panel actually reports the access code used to arm - for example on a PC1864:
```
259.75: 00100111 0 10011010 00000100 10010001 00000001 00000000 01010111 [0x27] Partition 1: Armed Bypass Trouble Backlight - Armed: Stay | Partition 2: Ready Trouble Backlight - Partition ready | Zones 1-8 open: none 
259.80: 00010110 0 00001110 01000010 10110001 00010111 [0x16] Panel version: v4.2 | Zone wiring: NC | Code length: 4 digits | *8 programming: no 
259.88: 10100101 0 00100010 01001000 10000010 00000000 10011100 00010100 01000001 [0xA5] 2022.02.04 02:00 | Partition 1 | Armed: Access code 4
```
In this log, the first panel command sent after arming is 0x27, and the panel status will get set to armed in `dsc.armed[partition]` - however, the access code in `dsc.accessCode[partition]` will still be the previously used access code.

After ~130ms, the panel sends the 0xA5 command with the access code used to arm - at that point, `dsc.accessCode[partition]` will be updated.  If the access code is different than the last used code, `doc.accessCodeChanged[partition]` will then be set to `true`.

Let me know if this time lag would cause an issue in your applications.  One way to handle this is to use `dsc.accessCodeChanged[partition]` to monitor when a different access code is being used on the partition.

Note that this is not an issue when disarming, the panel sends the disarm access code first and then sends the disarmed status, so `dsc.accessCode[partition]` will have the current disarm access code as soon as the panel status is marked as disarmed.

**Comment by taligentx** [comment]
The time lag where the arming access code is updated ~130ms after `armedChanged` is set is now also fixed in `develop` - the library no longer uses the 0x27 panel data to set `dsc.armedChanged[partition]`.  The library now waits until after the panel sends the arming access code and uses other commands that the panel sends later to set this flag and trigger the sketches.

This means that `dsc.accessCode[partition]` should now always have the currently used access code when checking `dsc.armedChanged[partition]` on both arming and disarming.

The changes are in the library code, there shouldn't be any code changes necessary in your sketches - let me know if this now works as expected.

**Comment by UltimatPronin** [comment]
@taligentx , thank you! Now access code detection used for arm/disarm work perfectly without any changes in my code.

**Reply by taligentx** [reply]
Great to hear, thanks for the feedback!  Included this fix in the new 3.0 release.
